Consume puya binary (#117)
algorandfoundation/puya-ts: docs/cli.md

### Puya

> Puya-ts is only the front end component of the compiler. After parsing and converting TypeScript into a common AST known as AWST, the remainder of the compilation is passed onto the Puya backend compiler. Puya is released as a binary from the Puya GitHub [repository](https://github.com/algorandfoundation/puya).

Reviewer: ```suggestion
Puya-ts is only the front end component of the compiler. After parsing and converting TypeScript into a common AST known as AWST, the remainder of the compilation is passed onto the Puya backend compiler. Puya is released as a binary on the Puya GitHub [repository](https://github.com/algorandfoundation/puya).
```

Currently, Puya binaries are available for:

- Linux (Intel and ARM)
- MacOS (Intel and Apple Silicon)
- Windows (Intel and ARM)

> If your system is supported, puya-ts will automatically download the appropriate binary for your system. It will make sure that the version of puya-ts and puya are compatible.

Reviewer: ```suggestion
If your system is supported, puya-ts will automatically download the appropriate binary for your system. 
```

### Puya (manual installation)

> This session is relevant if your platform is not supported by the Puya binaries or you need to run Puya in a specific way.

Reviewer: ```suggestion
This section is relevant if your platform is not supported by the Puya binaries ,or you need to run a custom version of Puya (eg. A pre-release version).
```

puya: error: the following arguments are required: --options, --awst
```

> After confirming that Puya is installed, you need to create a script file to trigger it. Then set the `PUYA_SCRIPT_PATH` environment variable to the path of the script file before running `puya-ts`. An example script file is shown below.

Reviewer: This seems unnecessarily complicated - can't we just use the previous approach of assuming requiring `puya` to be on the PATH? Or we could have a cli param for `--puya-path=puya` (would attempt to run puya from PATH) or `--puya-path=c:\custom\puya.exe` (run puya from specific path)

Using environment variables feels like an unnecessary complication and prevents a dev from easily having multiple versions available (eg. if they had one project using the current major version and another project using a beta, or historic version)

Author: The intention of the script path is so that the user can use `pipx run` to run a specified version of puya.
https://github.com/pypa/pipx?tab=readme-ov-file#walkthrough-running-an-application-in-a-temporary-virtual-environment
But I agree that it is complicated for an edge case. I've updated it to use a cli param like you suggested.

| `--target-avm-version`         | Integer                                                    | No (default current version) | Which AVM version to target for compilation output of contracts which do not explicitly identify their target version. Will default to current version in `mainnet` (with a small lag). Number must be a supported version of the compiler.                                                                                                                                                                                                          |
| `--cli-template-definition`    | KeyValuePair[]                                             | No                           | A list of template variable definitions which will be used to replace place holders in the compiled output. Values should be in the format MY_VAR_NAME=123 where the variable name does not include the template variable prefix. Binary values can be specified in hex strings using MY_VAR=0x1234FF                                                                                                                                                |

> | `--template-vars-prefix`       | String                                                     | No (default 'TMPL\_')        | The prefix used by template variables in the compilation targets.                                                                                                                                                                                                                                                                                                                                                                                    |

Author: This seems be prettier applied auto format. It adds `\` in front of `_` because it's an escape characters.
